File: src/infra/controllers/webhook/createWebhookController.ts
import { CreateWebhookUseCase } from "../../../app/useCases/webhook/createWebhookUseCase";
import { AuthMiddleware } from "../../../main/middlewares/authMiddleware";
import { RouteDTO } from "../../../main/types/RouteDTO";
import { ErrorHandlerAdapter } from "../../adapters/errorHandlerAdapter";
import { SchemaValidatorAdapter } from "../../adapters/schemaValidatorAdapter";
import { createWebhookSchema } from "../../schemas/internal/webhook";

class CreateWebhookController {
  constructor(private createWebhookUseCase: CreateWebhookUseCase) {}

  async handle(route: RouteDTO) {
    try {
      const { token } = await AuthMiddleware.authenticate(route);

      const body = route.request.body;
      const params = route.request.params;

      const schemaValidator = new SchemaValidatorAdapter(createWebhookSchema);
      const data = schemaValidator.validate({ ...body, ...params });

      const webhook = await this.createWebhookUseCase.execute(data, token);

      return route.response.json(webhook, 201);
    } catch (error) {
      return ErrorHandlerAdapter.handle(error);
    }
  }
}

export { CreateWebhookController };

File: internal/infra/controllers/webhook/createWebhookController.go
package webhookControllers

import (
	"encoding/json"
	"net/http"

	webhookUseCases "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/app/useCases/webhook"
	"github.com/go-playground/validator/v10"
)

type CreateWebhookController struct {
	createWebhookUseCase webhookUseCases.CreateWebhookUseCase
}

func (c *CreateWebhookController) Handle(w http.ResponseWriter, r *http.Request) error {
	var input webhookUseCases.CreateWebhookUseCaseInput

	decoder := json.NewDecoder(r.Body)
	decoder.DisallowUnknownFields()

	if error := decoder.Decode(&input); error != nil {
		return error
	}

	validate := validator.New()

	if error := validate.Struct(input); error != nil {
		return error
	}

	return nil
}
